### ISSUE TITLE
Message context

### ISSUE THREAD

**Issue by BrianVallelunga** [Azure/azure-webjobs-sdk#644]
This resurrects part of the issue at: https://github.com/Azure/azure-webjobs-sdk/issues/275

I've been trying to figure out how to use a custom child container that injects message-specific dependencies into a class with a ServiceBusTrigger handler on it.

My plan was to use the child container from within my custom IJobActivator to construct the trigger handler. Unfortunately, I don't have any way (that I can see) of getting any contextual information into my job activator. I thought that perhaps I could put context onto the thread, but that doesn't seem possible since everything is async and seems to use different threads.

The suggestion at the time was to use a custom MessageProcessor and MessageProvider. From my testing, that really only allows me to take over the whole processing pipeline instead of using what is already provided by the SDK with a small change.

Does anyone have any thoughts on how to accomplish this without throwing out most of the existing structure?

**Comment by BrianVallelunga** [comment]
According to this article: http://blog.stephencleary.com/2013/04/implicit-async-context-asynclocal.html and my own tests, I can use `CallContext.LogicalSetData` and `CallContext.LogicalGetData` to store a context associated with the message that is maintained across async calls. With this functionality, I can use a custom MessageProcessor to create and dispose of the child container and to store the message context initially. My custom JobActivator will be able to use the message context to get the child container and construct the tenant-specific handler class.

I'll post back when I have fully tested this, and will try to include some code for others needing a similar solution. If someone out there reading this knows of any reason not to do this, please speak up.

**Comment by DaveMiscampbell** [comment]
I'd be really interested to see what you come up with.

**Comment by BrianVallelunga** [comment]
I'm happy to report I have a working solution to maintaining and accessing a consistent message context from anywhere I need it in the execution of a service bus message. 

I've created a Gist of the basic code at: https://gist.github.com/BrianVallelunga/418e086d1758a4bd1441

The core of the functionality is in the MessageScopeManager  class which implements the following interface:

```c#
public interface IMessageScopeManager
{
    void BeginMessageScope(BrokeredMessage message);
    void CompleteMessageScope();
    IUnityContainer GetScopedContainer();
}
```

The manager is used by the MessageProcessor to in the corresponding BeginProcessingMessageAsync and CompleteProcessingMessageAsync methods.

The custom IJobActivator uses the GetScopedContainer() method to access a child container that is used to create job instances.

I'm open to any suggestions and improvements. I used a ConcurrentDictionary to hold the child containers in the MessageScopeManager and I'm not sure if that is the right approach or not.

Feel free to use the code however you want. There's no license associated with it.
